### ISSUE TITLE
Recompute steps_per_epoch in training_generator.py::fit_generator

### ISSUE THREAD

**Issue by vikua** [keras-team/keras#13038]
## Rationale 

Sometimes it can be useful to change value `batch_size` between epochs (https://arxiv.org/abs/1711.00489), but looks like there is no clean way to do so in Keras right now. 
Obvious approach would be to do a loop outside of `model.fit` and change batch_size at the end of iteration: 

```
batch_size = ... 
X, y = ... 
for i in range(epochs):
    model.fit(X, y, batch_size=batch_size)
    batch_size = ...  # change batch size here 
``` 

But this approach makes usage of existing Keras Callbacks (like `EarlyStopping`, `ReduceLearningRateOnPlateu`, etc.) impossible. 

As https://github.com/keras-team/keras/issues/10143 suggests it is possible to implement custom Sequence which changes batch size, but as @Dref360 says (https://github.com/keras-team/keras/issues/10143#issuecomment-448239696) number of steps per epoch will never be recomputed and each epoch will progressively increase.

I've looked into the code and found that it shouldn't be very hard to recompute number of `steps_per_epoch` after each epoch. 

If this issue makes sense, I've created a small PR  and looking for a feedback. 
https://github.com/keras-team/keras/pull/13037

**Comment by dabasajay** [comment]
Hi @vikua 
You're right that it shouldn't be very hard to recompute the number of `steps_per_epoch` after each epoch as you suggested in your PR https://github.com/keras-team/keras/pull/13037,
![1](https://user-images.githubusercontent.com/38384266/60464587-a6c77000-9c6c-11e9-9d39-84f0443694d6.png)

However, there are other things to consider with this change. 
(All changes are done in training_utils.py file)
1. Progress bar needs to be updated with recomputed `steps_per_epoch` after each epoch. I fixed it with this and it worked:
![2](https://user-images.githubusercontent.com/38384266/60464692-edb56580-9c6c-11e9-8e9b-ff851f23ffe9.png)
2. There are many scenarios but let's consider a simple one:
Say `fit_generator` is called as `fit_generator(gen,epochs=6,workers = 0)` where `gen` is an instance of `Sequence` (`keras.utils.Sequence`) object and with `workers=0`, the generator is executed on the main thread and now `gen` must implement its own `on_epoch_end`(or `on_epoch_start`) function in order to change batch_size during runtime. I fixed it with this and it worked: (Full code below)
![3](https://user-images.githubusercontent.com/38384266/60465447-bb0c6c80-9c6e-11e9-8389-d3076cc10c21.png)
3. Third and the problematic one, if you go for `workers>0` and since you now have an `on_epoch_end`(or `on_epoch_start`) function with `gen` then the above code will work **but** results are bit different this time. Now `gen` an instance of `OrderedEnqueuer` class which internally calls `on_epoch_end` function and so the function is called multiple times for the same epoch.
(In data_utils.py)
![4](https://user-images.githubusercontent.com/38384266/60466506-79c98c00-9c71-11e9-95b7-989c68987a47.png)

Code (Changes are done in training_utils.py according to the image shown in 2nd pt):
```python
import os
import tensorflow as tf
import numpy as np
from keras.models import Sequential
from keras.layers import Dense, Activation
from keras.utils import Sequence
tf.logging.set_verbosity(tf.logging.ERROR)
X = np.random.randn(600,10)
y = np.random.randint(0,2,600)
model = Sequential()
model.add(Dense(1,input_shape=(10,)))
model.add(Activation('sigmoid'))
model.compile(loss='binary_crossentropy',optimizer='adam',metrics=['accuracy'])
class BatchSizeScheduler(Sequence):
    def __init__(self, x_set, y_set, initial_batch_size, batch_scheduler=None):
        self.x, self.y = x_set, y_set
        self.batch_size = initial_batch_size
        self.batch_scheduler = batch_scheduler
    def __len__(self):
        return int(np.ceil(len(self.x) / float(self.batch_size)))
    def __getitem__(self, idx):
        batch_x = self.x[idx * self.batch_size:(idx + 1) * self.batch_size]
        batch_y = self.y[idx * self.batch_size:(idx + 1) * self.batch_size]
        return np.array(batch_x), np.array(batch_y)
    def on_epoch_end(self):
        if self.batch_scheduler is not None:
            self.batch_size = self.batch_scheduler(self.batch_size)
gen = BatchSizeScheduler(X,y,10,lambda x:x+10)
model.fit_generator(gen,epochs=6,workers = 0)
```
    Epoch 1/6
    60/60 [==============================] - 0s 5ms/step - loss: 0.9846 - acc: 0.4600
    Epoch 2/6
    30/30 [==============================] - 0s 1ms/step - loss: 0.9634 - acc: 0.4617
    Epoch 3/6
    20/20 [==============================] - 0s 2ms/step - loss: 0.9514 - acc: 0.4633
    Epoch 4/6
    15/15 [==============================] - 0s 2ms/step - loss: 0.9428 - acc: 0.4650
    Epoch 5/6
    12/12 [==============================] - 0s 2ms/step - loss: 0.9362 - acc: 0.4650
    Epoch 6/6
    10/10 [==============================] - 0s 2ms/step - loss: 0.9307 - acc: 0.4633

```python
model.fit_generator(gen,epochs=6)
```
    Epoch 1/6
    60/60 [==============================] - 0s 5ms/step - loss: 0.8638 - acc: 0.5033
    Epoch 2/6
    30/30 [==============================] - 0s 5ms/step - loss: nan - acc: nan
    Epoch 3/6
    15/15 [==============================] - 0s 1ms/step - loss: nan - acc: nan
    Epoch 4/6
    12/12 [==============================] - 0s 2ms/step - loss: nan - acc: nan  
    Epoch 5/6
    10/10 [==============================] - 0s 11ms/step - loss: nan - acc: nan
    Epoch 6/6
    8/8 [==============================] - 0s 2ms/step - loss: nan - acc: nan

Please suggest if there's a fix for this or if I'm doing something wrong.

**Comment by vikua** [comment]
Hi @dabasajay! Thanks you for your feedback, it is extremely helpful. 
I updated the PR to address your key points. Specifically: 

1. Progress bar - you are right, yes we definitely need to reset parameters once again. Good catch! Unfortunately callbacks doesn't purport parameters update, so as you suggested we need to reset all parameters from altogether. 

2 and 3 - I guess there is better way to approach it. We can't call `on_epoch_end` in training loop cause, as you said, `on_epoch_end` is also called in `OrderedEnqueuer` and we may end up with multiple calls to it. 
One simple way to do it would be to check if `generator` is instance of python generator or of `workers <= 0` and call `on_epoch_end` only in this case, but I found that if  `on_epoch_end` inside of `OrderedEnqueuer` then we probably should keep it that way for python generator in main thread. 
I've added `on_epoch_end` to `training_utils.py::iter_sequence_infinite` every time generator is exhausted (which means epoch has finished).

My PR is updated. 
Thanks a lot!

**Comment by vikua** [comment]
The only issue I'm getting - in some reasons 1st and 2nd epoch are always getting same batch size. 
It is true in my unit tests (same amount of batches for those epochs) and I cat reproduce it with 

```
import numpy as np
import keras

from keras import Model
from keras.layers import Dense, Input
from keras.utils import Sequence
from keras.datasets import boston_housing


class BatchSizeSchedulerSequence(Sequence):
    def __init__(self, X, y, initial_batch_size=32):
        self.X = X
        self.y = y
        self.batch_size = initial_batch_size

    def __len__(self):
        new_len = int(np.ceil(len(self.X) / float(self.batch_size)))
        # print('BSS Len:', new_len, 'batch_size', self.batch_size)
        return new_len

    def __getitem__(self, idx):
        batch_x = self.X[idx * self.batch_size:(idx + 1) * self.batch_size]
        batch_y = self.y[idx * self.batch_size:(idx + 1) * self.batch_size]
        return np.array(batch_x), np.array(batch_y)

    def on_epoch_end(self):
        self.batch_size += 2

(X_train, y_train), (X_test, y_test) = boston_housing.load_data()

inputs = Input(shape=(X_train.shape[1],))
outputs = Dense(64, activation='relu')(inputs)
outputs = Dense(1)(outputs)
model = Model(inputs, outputs)

model.compile(optimizer='adam', loss='mae')

train_seq = BatchSizeSchedulerSequence(X_train, y_train, initial_batch_size=2)
model.fit_generator(train_seq, workers=0, epochs=10, validation_data=(X_test, y_test))
```
Output is always 

```
Epoch 1/10
202/202 [==============================] - 0s 1ms/step - loss: 10.0856 - val_loss: 7.1507
Epoch 2/10
202/202 [==============================] - 0s 898us/step - loss: 5.1844 - val_loss: 5.4457
Epoch 3/10
51/51 [==============================] - 0s 894us/step - loss: 5.5938 - val_loss: 5.9979
Epoch 4/10
41/41 [==============================] - 0s 933us/step - loss: 5.1119 - val_loss: 4.9129
Epoch 5/10
34/34 [==============================] - 0s 917us/step - loss: 4.6283 - val_loss: 4.9017
Epoch 6/10
26/26 [==============================] - 0s 972us/step - loss: 4.4177 - val_loss: 4.9657
Epoch 7/10
23/23 [==============================] - 0s 946us/step - loss: 4.3916 - val_loss: 4.6934
Epoch 8/10
21/21 [==============================] - 0s 928us/step - loss: 4.2657 - val_loss: 4.7296
Epoch 9/10
19/19 [==============================] - 0s 977us/step - loss: 4.1856 - val_loss: 4.7973
Epoch 10/10
17/17 [==============================] - 0s 946us/step - loss: 4.1334 - val_loss: 4.5759
```

**Comment by dabasajay** [comment]
@vikua The issue is not just with the first epoch. As the `batch_size` is doubled every time starting with `2` and `number_of_samples = 404`, the sequence should've been `202, 101, 68, 51, 41, 34, 29, 26, 23, 21`. Some of them are missing.

**Comment by vikua** [comment]
I think I know what the problem is. 
Both `OrderedEnqueuer` and `workers==0` mode use python generators behind the scene. 
I'll try to explain in terms of `workers==0` mode. 
When generator emits value, it basically stops at `yield` operator. 

Generator `iter_sequence_infinite` looks like:  

```
def iter_sequence_infinite(seq):
    """Iterate indefinitely over a Sequence.

    # Arguments
        seq: Sequence object

    # Returns
        Generator yielding batches.
    """
    while True:
        for i, item in enumerate(seq):
            yield item
        seq.on_poch_end()
```
Initial assumption was that once it emits last value in `seq` it immediately calls `seq.on_poch_end()`. But of course it is not happening (stupid me!). 
Correct version of this function would be 

```
def iter_sequence_infinite(seq):
    """Iterate indefinitely over a Sequence.

    # Arguments
        seq: Sequence object

    # Returns
        Generator yielding batches.
    """
    while True:
        l = len(seq)
        for i, item in enumerate(seq):
            if i == l - 1:
                seq.on_poch_end()
            yield item
```

Here we guarantee that we call `seq.on_poch_end()` when last element is taken from `seq`, and steps_per_epochs are correct now: 

```
Epoch 1/10
202/202 [==============================] - 0s 2ms/step - loss: 9.5407 - val_loss: 7.9571
Epoch 2/10
101/101 [==============================] - 0s 963us/step - loss: 5.4546 - val_loss: 5.6195
Epoch 3/10
68/68 [==============================] - 0s 1ms/step - loss: 5.0326 - val_loss: 5.0873
Epoch 4/10
51/51 [==============================] - 0s 1ms/step - loss: 5.0866 - val_loss: 5.3378
Epoch 5/10
41/41 [==============================] - 0s 1ms/step - loss: 4.8442 - val_loss: 5.1842
Epoch 6/10
34/34 [==============================] - 0s 1ms/step - loss: 4.9167 - val_loss: 4.7993
Epoch 7/10
29/29 [==============================] - 0s 986us/step - loss: 4.4372 - val_loss: 4.6037
Epoch 8/10
26/26 [==============================] - 0s 1ms/step - loss: 4.2819 - val_loss: 4.7456
Epoch 9/10
23/23 [==============================] - 0s 1ms/step - loss: 4.3526 - val_loss: 4.7284
Epoch 10/10
21/21 [==============================] - 0s 1ms/step - loss: 4.3442 - val_loss: 4.4588
```

But we need different solution for `workers > 0` as well...

**Comment by dabasajay** [comment]
I think I found a fix for this. Check out my proposal in keras-user google groups here:
https://docs.google.com/document/d/1G5epNPN17MVuKvlAEM8suZmoXDI8Vv1DTXntlILljac/edit?usp=drive_web

**Comment by vikua** [comment]
@dabasajay Sorry, didn't notice you reply right away. Yeah, I also was thinking about it. 

I found another way to tackle this. I fixed the issue in my PR without changing the code too much. Now it works in every case.

**Comment by dabasajay** [comment]
Hey @vikua  
So my PR didn't pass the unit tests and when I looked into the tests, I couldn't figure out the cause. You're right and thanks for the information. I closed my PR as this approach is not gonna work and now I'll look into your current PR. I'll let you know if I find something buggy.
